code/obj/item/gun/reagent.dm

```diff
@@ -0,0 +1,146 @@
+/obj/item/gun/reagent
+	name = "reagent gun"
+	icon = 'icons/obj/items/gun.dmi'
+	item_state = "gun"
+	m_amt = 2000
+	g_amt = 1000
+	mats = 16
+	add_residue = 0 // Does this gun add gunshot residue when fired? Energy guns shouldn't.
+	var/capacity = 100 // reagent capacity of the gun
+	var/list/ammo_reagents = null // list of reagents accepted as ammo, leave blank if you want any to be accepted
+	var/projectile_reagents = 0 // whether the reagents should get transfered to the projectiles
+	var/dump_reagents_on_turf = 0 //set this to 1 if you want the dumped reagents to be put onto the turf instead of just evaporated into nothingness
+	var/custom_reject_message = "" //set this to a string if you want a custom message to be shown instead of the default when a reagent isnt accepted by the gun
+
+	New()
+		src.create_reagents(capacity)
+		..()
+
+	is_open_container()
+		return 1
+
+	alter_projectile(var/obj/projectile/P)
+		if(src.projectile_reagents && P && P.proj_data)
+			if (!P.reagents)
+				P.reagents = new /datum/reagents(P.proj_data.cost)
+				P.reagents.my_atom = P
+			src.reagents.trans_to(P, P.proj_data.cost)
+
+	on_reagent_change(add)
+		if(!add || !src.ammo_reagents)
+			src.update_icon()
+			return
+		var/mob/M = ismob(src.loc) ? src.loc : null
+		global.check_whitelist(src, src.ammo_reagents, M, src.custom_reject_message)
+		src.update_icon()
+
+	get_desc()
+		. = "[src.projectiles ? "It is set to [src.current_projectile.sname]. " : ""]There are [src.reagents.total_volume]/[src.reagents.maximum_volume] units left!"
+		if(src.current_projectile)
+			. += " Each shot will currently use [src.current_projectile.cost] units!"
+		else
+			. += "<span class='alert'>*ERROR* No output selected!</span>"
+		..()
+
+	update_icon()
+		return 0
+
+	canshoot()
+		if(src.reagents && src.current_projectile)
+			if(src.reagents.total_volume >= src.current_projectile.cost)
+				return 1
+		return 0
+
+	process_ammo(var/mob/user)
+		if (!canshoot())
+			boutput(user, "<span class='alert'>\The [src]'s internal reservoir does not contain enough reagents to fire it!</span>")
+		if(!src.projectile_reagents)
+			src.reagents.remove_any(src.current_projectile.cost)
+			src.update_icon()
+		return 1
+
+	MouseDrop(over_object, src_location, over_location)
+		..()
+		if(!isliving(usr))
+			return
+
+		if(get_dist(src, usr) > 1)
+			boutput(usr, "<span class='alert'>You need to be closer to empty \the [src] out!</span>")
+			return
+
+		if (!src.reagents)
+			boutput(usr, "<span class='alert'>The little cap on the fluid container is stuck. Uh oh.</span>")
+			return
+
+		if(src.reagents.total_volume)
+			if (src.dump_reagents_on_turf)
+				logTheThing("combat", usr, null, "transfers chemicals from [src] [log_reagents(src)] to [get_turf(src)] at [log_loc(usr)].")
+				src.reagents.trans_to(get_turf(src), src.reagents.total_volume)
```

Comment:
I'm not completely sure but I think this actually transfers the reagents *into* the turf rather than create a fluid. I think that 
```suggestion
				src.reagents.reaction(get_turf(src))
				src.reagents.clear_reagents()
```
might work?